{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import os.path\n",
    "import csv\n",
    "\n",
    "class PatientPhenotype:\n",
    "    \n",
    "    def __init__(self, eid, case, sex, yearBirth):\n",
    "        \n",
    "        self.eid = eid\n",
    "        self.case = case\n",
    "        self.sex = sex\n",
    "        self.yearBirth = yearBirth\n",
    "        self.snps = {}\n",
    "        \n",
    "    def getEid(self):\n",
    "        return self.eid\n",
    "     \n",
    "    def getCase(self):\n",
    "        return self.case\n",
    "    \n",
    "    def getSex(self):\n",
    "        return self.sex\n",
    "    \n",
    "    def getYearBirth(self):\n",
    "        return self.yearBirth\n",
    "        \n",
    "    def addSnps(self, snpId, allele1,allele2):\n",
    "        self.snps[snpId] = Snp(snpId,allele1,allele2)\n",
    "        \n",
    "    def snpCode(self,snpId,allele1,allele2):\n",
    "       \n",
    "        self.snps[snpId].setSnpCode(allele1,allele2)\n",
    "        \n",
    "    def getSnpCode(self,snpId):\n",
    "        return self.snps[snpId].getSnpCode()\n",
    "        \n",
    "    def getSize(self):\n",
    "        return len(self.snps)\n",
    "        \n",
    "        \n",
    "class Snp:\n",
    "    \n",
    "    def __init__(self,snpId,allele1,allele2):\n",
    "        \n",
    "        self.snpId = snpId\n",
    "        self.allele1 = allele1\n",
    "        self.allele2 = allele2\n",
    "        self.snpCode = -1\n",
    "        \n",
    "    def getId(self):\n",
    "        \n",
    "        return self.snpId\n",
    "        \n",
    "    def getAllele1(self):\n",
    "        \n",
    "        return self.allele1\n",
    "        \n",
    "    def getAllele2(self):\n",
    "        \n",
    "        return self.allele2\n",
    "        \n",
    "    def setSnpCode(self,allele1,allele2):\n",
    "      \n",
    "        if (self.allele1 == allele1 and self.allele2 == allele2) or (self.allele1 == allele2 and self.allele2 == allele1):\n",
    "            code = 2\n",
    "           \n",
    "        elif (self.allele1 == allele1 and self.allele2 != allele2) or (self.allele1 == allele2 and self.allele2 != allele1):\n",
    "            code = 1\n",
    "           \n",
    "        elif (self.allele1 != allele1 and self.allele2 == allele2) or (self.allele1 != allele2 and self.allele2 == allele1):\n",
    "            code = 1\n",
    "            \n",
    "        elif (self.allele1 != allele1 and self.allele2 != allele2) or (self.allele1 != allele2 and self.allele2 != allele1):\n",
    "            code = 0\n",
    "            \n",
    "        self.snpCode = code\n",
    "        \n",
    "    def getSnpCode(self):\n",
    "        \n",
    "        return self.snpCode\n",
    "    \n",
    "    \n",
    "class SaveDataSet():\n",
    "    def __init__(self,name,ids,data):\n",
    "        \n",
    "        self.name = name\n",
    "        self.ids = ids\n",
    "        self.data = data\n",
    "        self.__writeLogFile()\n",
    "        self.__writeVariables('patients')\n",
    "        self.__writeVariables('snps')\n",
    "        self.__writeCSV()\n",
    "        \n",
    "    def __writeLogFile(self):\n",
    "        \n",
    "        write = open(self.name+'.log','w')\n",
    "        write.write(\"Log file for the Test Data Set \"+self.name+'\\n')\n",
    "        write.write(self.name + \" has \"+ str(len(self.ids['patients']['nameToId'].keys())) + \" patients \"'\\n')\n",
    "        write.write(self.name + \" has \"+ str(len(self.ids['snps']['nameToId'].keys())) + \" snps \"'\\n')\n",
    "        write.close()\n",
    "        \n",
    "    def __writeVariables(self,kind):\n",
    "        \n",
    "        write = open(self.name+'_'+kind+'.txt','w')\n",
    "        write.write(kind + '\\n')\n",
    "        \n",
    "        for i in range(len(self.ids[kind]['nameToId'].keys())):\n",
    "            name = self.ids[kind]['idToName'][i]\n",
    "            write.write(name + '\\t'+str(self.ids[kind]['nameToId'][name]) + '\\n')\n",
    "        \n",
    "        write.close()\n",
    "        \n",
    "        \n",
    "    def __writeCSV(self):\n",
    "        \n",
    "        f = open(self.name+'.csv', 'wt')\n",
    "        try:\n",
    "            writer = csv.writer(f)\n",
    "            \n",
    "            line = ' '\n",
    "            for i in range(len(self.ids['snps']['nameToId'].keys())):\n",
    "                if i == 0 :\n",
    "                    line += self.ids['snps']['idToName'][i]\n",
    "                else:\n",
    "                    line += ','+self.ids['snps']['idToName'][i]\n",
    "            writer.writerow([line])\n",
    "            \n",
    "            line = ' ' \n",
    "            for i in range(len(self.ids['snps']['nameToId'].keys())):\n",
    "                name = self.ids['snps']['idToName'][i]\n",
    "                if i == 0:\n",
    "                    line += str(self.ids['snps']['nameToId'][name])\n",
    "                else:\n",
    "                    line += ','+str(self.ids['snps']['nameToId'][name])\n",
    "            writer.writerow([line])\n",
    "            \n",
    "            \n",
    "            for i in range(len(self.ids['patients']['nameToId'].keys())):\n",
    "                line = ''\n",
    "                for j in range(len(self.ids['snps']['nameToId'].keys())):\n",
    "                    if j == 0:\n",
    "                        line += str(self.data[i,j]) \n",
    "                    else:\n",
    "                        \n",
    "                        line += ','+str(self.data[i,j])\n",
    "                        \n",
    "                writer.writerow([line])\n",
    "        finally:\n",
    "            f.close()\n",
    "        \n",
    "\n",
    "def setIdToName(aList):\n",
    "    \n",
    "    ids = {}\n",
    "    nameToId = {}\n",
    "    idToName = {}\n",
    "    count = 0\n",
    "    \n",
    "    for i in aList:\n",
    "        \n",
    "        nameToId[i] = count\n",
    "        idToName[count] = i\n",
    "        count += 1\n",
    "        \n",
    "    ids['nameToId'] = nameToId\n",
    "    ids['idToName'] = idToName\n",
    "    \n",
    "    return ids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "minorSnps = []\n",
    "chromosomes = {}\n",
    "\n",
    "for i in range(14):\n",
    "    \n",
    "    chro = 'chr'+str(i+1)\n",
    "    path = 'C:\\\\Users\\\\ANTONIS\\\\Desktop\\\\test1\\\\'+chro+'.assoc.fisher'\n",
    "    f = open(path,'r')\n",
    "    f.readline()\n",
    "    snps = {}\n",
    "    \n",
    "    for line in f:\n",
    "        alleles = []\n",
    "        minorSnps.append(line.split()[1])\n",
    "        alleles.append(line.split()[3])\n",
    "        alleles.append(line.split()[6])\n",
    "        snps[line.split()[1]] = alleles\n",
    "        \n",
    "    chromosomes[chro] = snps\n",
    "    \n",
    "for i in range(14):\n",
    "    \n",
    "    chro = 'chr'+str(i+1)\n",
    "    path = 'C:\\\\Users\\\\ANTONIS\\\\Desktop\\\\test1\\\\'+chro+'snpList.txt'\n",
    "    write = open(path,'w')\n",
    "    for snp in chromosomes[chro].keys():\n",
    "        write.write(snp + '\\n')\n",
    "            \n",
    "    write.close()\n",
    "    \n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# TRAIN SET CREATION"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "########   TRAIN ##########\n",
    "\n",
    "patientsPhen = {}\n",
    "\n",
    "f = open('C:\\\\Users\\\\Antonis\\\\Documents\\\\GitHub\\\\diplwmatikh\\\\data\\\\phenotype_euro_train.txt','r')\n",
    "f.readline()\n",
    "\n",
    "for line in f:\n",
    "    patientsPhen[line.split()[0]] = PatientPhenotype(line.split()[0],line.split()[3],line.split()[1],line.split()[2])\n",
    "\n",
    "    \n",
    "path = 'C:\\\\Users\\\\ANTONIS\\\\Desktop\\\\test1\\\\'+'trainPatient.txt'\n",
    "write = open(path,'w')\n",
    "for patient in patientsPhen.keys():\n",
    "    write.write(patient + '\\n')\n",
    "            \n",
    "write.close()\n",
    "    \n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# run train_lgen bat"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "count is  4482\n",
      "snps is  334\n",
      "shape id  (4482, 334)\n",
      "xtraining =  4482\n",
      "xtraining.T =  334\n"
     ]
    }
   ],
   "source": [
    "ids = {}\n",
    "for i in range(14):\n",
    "    \n",
    "    chro = 'chr'+str(i+1)\n",
    "    path = 'C:\\\\Users\\\\ANTONIS\\\\Desktop\\\\test1\\\\'+chro+'train.lgen'\n",
    "    \n",
    "    if os.path.exists(path):\n",
    "        f = open(path,'r')\n",
    "    \n",
    "        for line in f:\n",
    "            patientsPhen[line.split()[0]].addSnps(line.split()[2], line.split()[3],line.split()[4])\n",
    "    \n",
    "\n",
    "    \n",
    "for i in range(14):\n",
    "    \n",
    "    chro = 'chr'+str(i+1)\n",
    "    \n",
    "    for snp in chromosomes[chro].keys():\n",
    "        \n",
    "        allele1 = chromosomes[chro][snp][0]\n",
    "        allele2 = chromosomes[chro][snp][1]\n",
    "    \n",
    "        for patient in patientsPhen.keys():\n",
    "            \n",
    "            patientsPhen[patient].snpCode(snp,allele1,allele2)\n",
    "              \n",
    "count = 0\n",
    "\n",
    "for patient in patientsPhen.keys():\n",
    "            \n",
    "    if patientsPhen[patient].getSize() == len(minorSnps):\n",
    "        \n",
    "        count += 1\n",
    "        \n",
    "print(\"count is \",count)\n",
    "print(\"snps is \",len(minorSnps))\n",
    "\n",
    "\n",
    "ids['patients'] = setIdToName(list(patientsPhen.keys()))\n",
    "ids['snps'] = setIdToName(minorSnps)\n",
    "    \n",
    "xTraining = np.zeros((len(ids['patients']['nameToId'].keys()), len(ids['snps']['nameToId'].keys())),dtype = int)\n",
    "\n",
    "print(\"shape id \",xTraining.shape)\n",
    "print(\"xtraining = \",len(xTraining))\n",
    "print(\"xtraining.T = \",len(xTraining.T))\n",
    "    \n",
    "for i in range(len(xTraining)):\n",
    "    for j in range(len(xTraining.T)):\n",
    "        \n",
    "        patient = ids['patients']['idToName'][i]\n",
    "        snp = ids['snps']['idToName'][j]\n",
    "        \n",
    "        xTraining[i,j] = patientsPhen[patient].getSnpCode(snp)\n",
    "        \n",
    "        \n",
    "yTraining = []\n",
    "\n",
    "for i in range(len(ids['patients']['nameToId'].keys())):\n",
    "    \n",
    "    patient = ids['patients']['idToName'][i]\n",
    "    yTraining.append(patientsPhen[patient].getCase())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-56-c5ca800c0c74>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mdok\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mSaveDataSet\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;34m\"dok\"\u001b[0m\u001b[1;33m,\u001b[0m\u001b[0mids\u001b[0m\u001b[1;33m,\u001b[0m\u001b[0mxTraining\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;32m<ipython-input-50-47c7ba9441d9>\u001b[0m in \u001b[0;36m__init__\u001b[0;34m(self, name, ids, data)\u001b[0m\n\u001b[1;32m     90\u001b[0m         \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0m__writeVariables\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;34m'patients'\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m     91\u001b[0m         \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0m__writeVariables\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;34m'snps'\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m---> 92\u001b[0;31m         \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0m__writeCSV\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     93\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m     94\u001b[0m     \u001b[1;32mdef\u001b[0m \u001b[0m__writeLogFile\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mself\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m<ipython-input-50-47c7ba9441d9>\u001b[0m in \u001b[0;36m__writeCSV\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    143\u001b[0m                         \u001b[0mline\u001b[0m \u001b[1;33m+=\u001b[0m \u001b[1;34m','\u001b[0m\u001b[1;33m+\u001b[0m\u001b[0mstr\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mdata\u001b[0m\u001b[1;33m[\u001b[0m\u001b[0mi\u001b[0m\u001b[1;33m,\u001b[0m\u001b[0mj\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m    144\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m--> 145\u001b[0;31m                 \u001b[0mwriter\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mwriterow\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m[\u001b[0m\u001b[0mline\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    146\u001b[0m         \u001b[1;32mfinally\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m    147\u001b[0m             \u001b[0mf\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mclose\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "dok = SaveDataSet(\"dok\",ids,xTraining)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "ename": "CParserError",
     "evalue": "Error tokenizing data. C error: out of memory",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mCParserError\u001b[0m                              Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-57-59cebdfaf00c>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[1;32mimport\u001b[0m \u001b[0mpandas\u001b[0m \u001b[1;32mas\u001b[0m \u001b[0mpd\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m----> 2\u001b[0;31m \u001b[0mdf\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mpd\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mread_csv\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;34m'dok.csv'\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      3\u001b[0m \u001b[0mdf\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[0;32mC:\\Users\\ANTONIS\\Anaconda3\\lib\\site-packages\\pandas\\io\\parsers.py\u001b[0m in \u001b[0;36mparser_f\u001b[0;34m(filepath_or_buffer, sep, delimiter, header, names, index_col, usecols, squeeze, prefix, mangle_dupe_cols, dtype, engine, converters, true_values, false_values, skipinitialspace, skiprows, skipfooter, nrows, na_values, keep_default_na, na_filter, verbose, skip_blank_lines, parse_dates, infer_datetime_format, keep_date_col, date_parser, dayfirst, iterator, chunksize, compression, thousands, decimal, lineterminator, quotechar, quoting, escapechar, comment, encoding, dialect, tupleize_cols, error_bad_lines, warn_bad_lines, skip_footer, doublequote, delim_whitespace, as_recarray, compact_ints, use_unsigned, low_memory, buffer_lines, memory_map, float_precision)\u001b[0m\n\u001b[1;32m    560\u001b[0m                     skip_blank_lines=skip_blank_lines)\n\u001b[1;32m    561\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m--> 562\u001b[0;31m         \u001b[1;32mreturn\u001b[0m \u001b[0m_read\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mfilepath_or_buffer\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mkwds\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    563\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m    564\u001b[0m     \u001b[0mparser_f\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0m__name__\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mname\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[0;32mC:\\Users\\ANTONIS\\Anaconda3\\lib\\site-packages\\pandas\\io\\parsers.py\u001b[0m in \u001b[0;36m_read\u001b[0;34m(filepath_or_buffer, kwds)\u001b[0m\n\u001b[1;32m    323\u001b[0m         \u001b[1;32mreturn\u001b[0m \u001b[0mparser\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m    324\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m--> 325\u001b[0;31m     \u001b[1;32mreturn\u001b[0m \u001b[0mparser\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mread\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    326\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m    327\u001b[0m _parser_defaults = {\n",
      "\u001b[0;32mC:\\Users\\ANTONIS\\Anaconda3\\lib\\site-packages\\pandas\\io\\parsers.py\u001b[0m in \u001b[0;36mread\u001b[0;34m(self, nrows)\u001b[0m\n\u001b[1;32m    813\u001b[0m                 \u001b[1;32mraise\u001b[0m \u001b[0mValueError\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;34m'skip_footer not supported for iteration'\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m    814\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m--> 815\u001b[0;31m         \u001b[0mret\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0m_engine\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mread\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mnrows\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    816\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m    817\u001b[0m         \u001b[1;32mif\u001b[0m \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0moptions\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mget\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;34m'as_recarray'\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[0;32mC:\\Users\\ANTONIS\\Anaconda3\\lib\\site-packages\\pandas\\io\\parsers.py\u001b[0m in \u001b[0;36mread\u001b[0;34m(self, nrows)\u001b[0m\n\u001b[1;32m   1312\u001b[0m     \u001b[1;32mdef\u001b[0m \u001b[0mread\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mself\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mnrows\u001b[0m\u001b[1;33m=\u001b[0m\u001b[1;32mNone\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m   1313\u001b[0m         \u001b[1;32mtry\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1314\u001b[0;31m             \u001b[0mdata\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0m_reader\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mread\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mnrows\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1315\u001b[0m         \u001b[1;32mexcept\u001b[0m \u001b[0mStopIteration\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m   1316\u001b[0m             \u001b[1;32mif\u001b[0m \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0m_first_chunk\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[0;32mpandas\\parser.pyx\u001b[0m in \u001b[0;36mpandas.parser.TextReader.read (pandas\\parser.c:8748)\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;32mpandas\\parser.pyx\u001b[0m in \u001b[0;36mpandas.parser.TextReader._read_low_memory (pandas\\parser.c:9003)\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;32mpandas\\parser.pyx\u001b[0m in \u001b[0;36mpandas.parser.TextReader._read_rows (pandas\\parser.c:9731)\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;32mpandas\\parser.pyx\u001b[0m in \u001b[0;36mpandas.parser.TextReader._tokenize_rows (pandas\\parser.c:9602)\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;32mpandas\\parser.pyx\u001b[0m in \u001b[0;36mpandas.parser.raise_parser_error (pandas\\parser.c:23325)\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;31mCParserError\u001b[0m: Error tokenizing data. C error: out of memory"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "df = pd.read_csv('dok.csv')\n",
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# TEST SET CREATION"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "patientsPhenTest = {}\n",
    "\n",
    "f = open('C:\\\\Users\\\\Antonis\\\\Documents\\\\GitHub\\\\diplwmatikh\\\\data\\\\phenotype_euro_test.txt','r')\n",
    "f.readline()\n",
    "for line in f:\n",
    "    patientsPhenTest[line.split()[0]] = PatientPhenotype(line.split()[0],line.split()[3],line.split()[1],line.split()[2])  \n",
    "    \n",
    "    \n",
    "path = 'C:\\\\Users\\\\ANTONIS\\\\Desktop\\\\test1\\\\'+'testPatient.txt'\n",
    "write = open(path,'w'):\n",
    "    for patient in patientsPhenTest.keys():\n",
    "        write.write(patient + '\\n')\n",
    "            \n",
    "write.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# run test_lgen bat"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "idsTest = {}\n",
    "\n",
    "for i in range(14):\n",
    "    \n",
    "    chro = 'chr'+str(i+1)\n",
    "    path = 'C:\\\\Users\\\\ANTONIS\\\\Desktop\\\\test1\\\\'+chro+'test.lgen'\n",
    "    if os.path.exists(path):\n",
    "        f = open(path,'r')\n",
    "    \n",
    "        for line in f:\n",
    "            patientsPhenTest[line.split()[0]].addSnps(line.split()[2], line.split()[3],line.split()[4]):\n",
    "    \n",
    "\n",
    "    \n",
    "for i in range(14):\n",
    "    \n",
    "    chro = 'chr'+str(i+1)\n",
    "    \n",
    "    for snp in chromosomes[chro].keys():\n",
    "        \n",
    "        allele1 = chromosomes[chro][snp][0]\n",
    "        allele2 = chromosomes[chro][snp][1]\n",
    "        \n",
    "        for patient in patientsPhenTest.keys():\n",
    "            \n",
    "            patientsPhenTest[patient].snpCode(snp,allele1,allele2)\n",
    "            \n",
    "            \n",
    "\n",
    "    \n",
    "count = 0\n",
    "\n",
    "for patient in patientsPhenTest.keys():\n",
    "            \n",
    "    if patientsPhenTest[patient].getSize() == len(minorSnps):\n",
    "        \n",
    "        count += 1\n",
    "        \n",
    "print(\"count is Test\",count)\n",
    "print(\"snps is Test\",len(minorsSnps))\n",
    "\n",
    "\n",
    "idsTest['patients'] = setIdToName(list(patientsPhenTest.keys()))\n",
    "idsTest['snps'] = setIdToName(minorSnps)\n",
    "    \n",
    "xTest = np.zeros((len(idsTest['patients']['nameToId'].keys()), ldtype=ype=(idsTest['snps']['nameToId'].keys())),dtype = int)\n",
    "\n",
    "print(\"shape id Test\",xTest.shape)\n",
    "print(\"xtraining = \",len(xTest))\n",
    "print(\"xtest.T = \",len(xTest.T))\n",
    "    \n",
    "for i in range(xTest):\n",
    "    for j in range(xTest.T):\n",
    "        \n",
    "        patient = idsTest['patient']['idToName'][i]\n",
    "        snp = idsTest['snp']['idToName'][j]\n",
    "        \n",
    "        xTest[i,j] = patientsPhenTest[patient].getSnpCode(snp)\n",
    "        \n",
    "        \n",
    "yTest = []\n",
    "\n",
    "for i in range(len(idsTest['patients']['nameToId'].keys()):\n",
    "    \n",
    "    patient = idsTest['patient']['idToName'][i]\n",
    "    yTest.append(patientsPhenTest[patient].getCase())\n",
    "\n",
    "yTest = np.array(yTest)"
   ]
  }
 ],
 "metadata": {
  "anaconda-cloud": {},
  "kernelspec": {
   "display_name": "Python [conda root]",
   "language": "python",
   "name": "conda-root-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
