{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "from sklearn import linear_model\n",
    "from sklearn.model_selection import cross_val_score\n",
    "from sklearn.metrics import mean_squared_error\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "import os.path\n",
    "import csv\n",
    "from sklearn.svm import SVC\n",
    "from sklearn.tree import DecisionTreeRegressor\n",
    "from sklearn.ensemble import RandomForestRegressor\n",
    "from sklearn.ensemble import RandomForestClassifier\n",
    "from sklearn import grid_search\n",
    "import sklearn.linear_model as linear_model\n",
    "import sklearn.metrics as metrics\n",
    "from sklearn import preprocessing\n",
    "import statsmodels.api as sm\n",
    "import statsmodels.formula.api as smf\n",
    "\n",
    "class PatientPhenotype:\n",
    "    \n",
    "    def __init__(self, eid, case, sex, yearBirth):\n",
    "        \n",
    "        self.eid = eid.strip()\n",
    "        self.case = int (case.strip())\n",
    "        self.sex = sex.strip()\n",
    "        self.yearBirth = yearBirth.strip()\n",
    "        self.snps = {}\n",
    "        \n",
    "    def getEid(self):\n",
    "        return self.eid\n",
    "     \n",
    "    def getCase(self):\n",
    "        return self.case\n",
    "    \n",
    "    def getSex(self):\n",
    "        return self.sex\n",
    "    \n",
    "    def getYearBirth(self):\n",
    "        return self.yearBirth\n",
    "        \n",
    "    def addSnps(self, snpId, allele1,allele2):\n",
    "        self.snps[snpId] = Snp(snpId,allele1,allele2)\n",
    "        \n",
    "    def snpCode(self,chromosomes = {}, snp = '', code = -1):\n",
    "    \n",
    "        if len(chromosomes.keys()) > 0:\n",
    "    \n",
    "             for i in range(len(chromosomes.keys())):\n",
    "    \n",
    "                chro = 'chr'+str(i+1)\n",
    "            \n",
    "                for snp in chromosomes[chro].keys():\n",
    "                \n",
    "                    allele1 = chromosomes[chro][snp][0].strip()\n",
    "                    allele2 = chromosomes[chro][snp][1].strip()\n",
    "    \n",
    "                    self.snps[snp.strip()].setSnpCode(allele1,allele2)\n",
    "        else:\n",
    "            \n",
    "            self.snps[snp.strip()].setCode(code)\n",
    "            \n",
    "    def getSnpCode(self,snpId):\n",
    "        return self.snps[snpId].getSnpCode()\n",
    "    \n",
    "    def getAllele1(self,snpId):\n",
    "        return self.snps[snpId].getAllele1()\n",
    "    \n",
    "    def getAllele2(self,snpId):\n",
    "        return self.snps[snpId].getAllele2()\n",
    "        \n",
    "    def getSize(self):\n",
    "        return len(self.snps)\n",
    "        \n",
    "        \n",
    "class Snp:\n",
    "    \n",
    "    def __init__(self,snpId,allele1,allele2):\n",
    "        \n",
    "        self.snpId = snpId\n",
    "        self.allele1 = allele1\n",
    "        self.allele2 = allele2\n",
    "        self.snpCode = -1\n",
    "        \n",
    "    def getId(self):\n",
    "        \n",
    "        return self.snpId\n",
    "        \n",
    "    def getAllele1(self):\n",
    "        \n",
    "        return self.allele1\n",
    "        \n",
    "    def getAllele2(self):\n",
    "        \n",
    "        return self.allele2\n",
    "        \n",
    "    def setSnpCode(self,allele1,allele2):\n",
    "      \n",
    "        if self.allele1.strip() == allele1.strip() and self.allele2.strip() == allele1.strip():\n",
    "            code = 2\n",
    "           \n",
    "        elif self.allele1.strip() == allele1.strip() and self.allele2.strip() != allele1.strip():\n",
    "            code = 1\n",
    "           \n",
    "        elif self.allele1.strip() != allele1.strip() and self.allele2.strip() == allele1.strip():\n",
    "            code = 1\n",
    "            \n",
    "        elif self.allele1.strip() != allele1.strip() and self.allele2.strip() != allele1.strip():\n",
    "            code = 0\n",
    "            \n",
    "        self.snpCode = code\n",
    "        \n",
    "    def setCode(self,aCode):\n",
    "        \n",
    "        self.snpCode = aCode\n",
    "        \n",
    "    def getSnpCode(self):\n",
    "        \n",
    "        return self.snpCode\n",
    "    \n",
    "class Read:\n",
    "    def __init__(self,path,numberOfChromosomes):\n",
    "        \n",
    "        self.chromosomes = {}\n",
    "        self.numberOfSnps = 0\n",
    "        self.path = path\n",
    "        self.numberOfChromosomes = numberOfChromosomes\n",
    "        \n",
    "    def readPatients(self,kind):\n",
    "        patients = {}\n",
    "        f = open(self.path + kind,'r')\n",
    "        f.readline()\n",
    "\n",
    "        for line in f:\n",
    "            patients[line.split()[0].strip()] = PatientPhenotype(line.split()[0],line.split()[3],line.split()[1],line.split()[2])\n",
    "        \n",
    "        f.close()\n",
    "        \n",
    "        return patients\n",
    "    \n",
    "    def readSnps(self,fileKind):\n",
    "        \n",
    "        for i in range(self.numberOfChromosomes):\n",
    "    \n",
    "            chro = 'chr'+str(i+1)\n",
    "            path = self.path + chro + fileKind\n",
    "            f = open(path,'r')\n",
    "            f.readline()\n",
    "    \n",
    "            self.chromosomes[chro] = self.__readSnpsOfChromosome(f)\n",
    "        \n",
    "            f.close()\n",
    "        \n",
    "        return self.chromosomes\n",
    "    \n",
    "    def __readSnpsOfChromosome(self,file):\n",
    "        \n",
    "        snps = {} \n",
    "       \n",
    "        for line in file:\n",
    "            \n",
    "            alleles = []\n",
    "            alleles.append(line.split()[3].strip())\n",
    "            alleles.append(line.split()[6].strip())\n",
    "            \n",
    "            snps[line.split()[1].strip()] = alleles\n",
    "            self.numberOfSnps += 1\n",
    "                \n",
    "        return snps\n",
    "        \n",
    "    def readLgen(self,patients,kind = ''):\n",
    "    \n",
    "        for i in range(self.numberOfChromosomes):\n",
    "            \n",
    "            chro = 'chr'+str(i+1)\n",
    "            path = self.path + chro + kind +'.lgen'\n",
    "    \n",
    "            if os.path.exists(path):\n",
    "                \n",
    "                f = open(path,'r')\n",
    "              \n",
    "                for line in f:\n",
    "                    \n",
    "                    if line.split()[0].strip() in patients.keys():\n",
    "                    \n",
    "                        patients[line.split()[0].strip()].addSnps(line.split()[2].strip(),line.split()[3].strip(),\n",
    "                                                                                              line.split()[4].strip())\n",
    "                f.close()\n",
    "       \n",
    "        return patients\n",
    "    \n",
    "    def getListOfSnps(self):\n",
    "        snps = []\n",
    "        for i in range(self.numberOfChromosomes):\n",
    "            chro = 'chr'+str(i+1)\n",
    "            for snp in self.chromosomes[chro].keys():\n",
    "                snps.append(snp)\n",
    "        \n",
    "        return snps\n",
    "        \n",
    "    def getNumberOfSnps(self):\n",
    "        \n",
    "        return self.numberOfSnps\n",
    "    \n",
    "    def readIds(self,kind):\n",
    "        \n",
    "        ids = {}\n",
    "        \n",
    "        ids['patients'] = self.__readPatientsId(kind)\n",
    "        ids['snps'] = self.__readSnpsId()\n",
    "        \n",
    "        return ids\n",
    "    \n",
    "    def __readPatientsId(self,kind):\n",
    "        \n",
    "        ids = {}\n",
    "        idToName = {}\n",
    "        nameToId = {}\n",
    "        \n",
    "        read = open(self.path + kind + 'Ids.txt','r')    \n",
    "        read.readline()\n",
    "        \n",
    "        for line in read:\n",
    "            \n",
    "            name = line.split('\\t')[0].strip()\n",
    "            idi = int (line.split('\\t')[1].strip())\n",
    "            \n",
    "            nameToId[name] = idi\n",
    "            idToName[idi] = name\n",
    "            \n",
    "        ids['nameToId'] = nameToId\n",
    "        ids['idToName'] = idToName\n",
    "        \n",
    "        read.close()\n",
    "                      \n",
    "        return ids  \n",
    "    \n",
    "    \n",
    "    def __readSnpsId(self):\n",
    "        \n",
    "        ids = {}\n",
    "        idToName = {}\n",
    "        nameToId = {}\n",
    "        \n",
    "        read = open(self.path + 'SnpsIds.txt','r')    \n",
    "        read.readline()\n",
    "        \n",
    "        for line in read:\n",
    "            \n",
    "            name = line.split('\\t')[0].strip()\n",
    "            idi = int (line.split('\\t')[1].strip())\n",
    "            \n",
    "            nameToId[name] = idi\n",
    "            idToName[idi] = name \n",
    "                      \n",
    "        ids['nameToId'] = nameToId\n",
    "        ids['idToName'] = idToName\n",
    "\n",
    "        read.close()\n",
    "                      \n",
    "        return ids  \n",
    "    \n",
    "    def readSnpsCode(self,patients,kind = ''):\n",
    "        \n",
    "        read = open(self.path + kind + 'SnpCode.txt','r')\n",
    "        read.readline()\n",
    "        read.readline()\n",
    "       \n",
    "        for line in read:   \n",
    "        \n",
    "            patient = line.split('\\t')[0].strip()\n",
    "            snp = line.split('\\t')[1].strip()\n",
    "            code = int (line.split('\\t')[2].strip())\n",
    "            allele1 = line.split('\\t')[3].strip()\n",
    "            allele2 = line.split('\\t')[4].strip()\n",
    "            if patient in patients.keys():\n",
    "                patients[patient].addSnps(snp,allele1,allele2)\n",
    "                patients[patient].snpCode(snp = snp,code = code)\n",
    "         \n",
    "        read.close()\n",
    "            \n",
    "        return patients\n",
    "            \n",
    "        \n",
    "    \n",
    "class Write:\n",
    "    \n",
    "    def __init__(self,path,numberOfChromosomes):\n",
    "        \n",
    "        self.path = path\n",
    "        self.numberOfChromosomes = numberOfChromosomes\n",
    "        \n",
    "    def writePatientsList(self,patients,kind):\n",
    "        \n",
    "        path = self.path + kind\n",
    "        write = open(path,'w')\n",
    "        for patient in patients.keys():\n",
    "            write.write(patient.strip() + '\\n')\n",
    "            \n",
    "        write.close()\n",
    "        \n",
    "        \n",
    "    def writeSnpsList(self,chromosomes):\n",
    "        \n",
    "        for i in range(self.numberOfChromosomes):\n",
    "    \n",
    "            chro = 'chr'+str(i+1)\n",
    "            path = self.path + chro + 'snpList.txt'\n",
    "            write = open(path,'w')\n",
    "            \n",
    "            for snp in chromosomes[chro].keys():\n",
    "                write.write(snp.strip() + '\\n')\n",
    "            \n",
    "            write.close()\n",
    "            \n",
    "    def saveData(self,ids,patients,patientKind,data,chroms = {}):\n",
    "        \n",
    "        self.__patientsLogFile(ids['patients'],patientKind)\n",
    "        self.__snpsLogFile(ids['snps'],chroms)\n",
    "        self.__snpCodeLog(ids['patients']['idToName'],ids['snps']['idToName'],patients,data,patientKind)\n",
    "        \n",
    "    def __patientsLogFile(self,ids,patientKind):\n",
    "        \n",
    "        write = open(self.path + patientKind + 'Ids.txt','w')\n",
    "        \n",
    "        write.write(str(len(ids['nameToId'])) + '\\n')\n",
    "        \n",
    "        for patient in ids['nameToId'].keys():\n",
    "            \n",
    "            write.write(patient.strip() + '\\t' + str(ids['nameToId'][patient]).strip() + '\\n')\n",
    "            \n",
    "        write.close()\n",
    "        \n",
    "    def __snpsLogFile(self,ids,chroms):\n",
    "        \n",
    "        if len(chroms.keys()) > 0:\n",
    "        \n",
    "            write = open(self.path + 'SnpsIds.txt','w')\n",
    "        \n",
    "            write.write(str(len(ids['nameToId'])) + '\\n')\n",
    "        \n",
    "            for chro in chroms.keys():\n",
    "              \n",
    "                for snp in chroms[chro].keys():\n",
    "                    write.write(snp.strip() + '\\t' + str(ids['nameToId'][snp.strip()]).strip() + '\\n')\n",
    "            \n",
    "            write.close()\n",
    "            \n",
    "    def __snpCodeLog(self,patientsIds,snpsIds,patients,data,patientKind):\n",
    "        \n",
    "        write = open(self.path + patientKind + 'SnpCode.txt','w')\n",
    "        \n",
    "        write.write(str(len(patientsIds)) + '\\n')\n",
    "        write.write(str(len(snpsIds)) + '\\n')\n",
    "        \n",
    "        for i in range(len(data)):\n",
    "            for j in range(len(data.T)):\n",
    "                allele1 = patients[patientsIds[i]].getAllele1(snpsIds[j])\n",
    "                allele2 = patients[patientsIds[i]].getAllele2(snpsIds[j])\n",
    "                write.write(patientsIds[i].strip() + '\\t' + snpsIds[j].strip() + '\\t' + str(data[i,j]).strip() + '\\t' \n",
    "                                                                            + allele1.strip() + '\\t' + allele2.strip() + '\\n')\n",
    "                \n",
    "        write.close()\n",
    "        \n",
    "        \n",
    "            \n",
    "class DataSet:\n",
    "    \n",
    "    def __init__(self,patients,ids):\n",
    "        \n",
    "        self.n = len(ids['patients']['nameToId'].keys())\n",
    "        self.m =len(ids['snps']['nameToId'].keys()) \n",
    "        self.patients = patients\n",
    "        self.ids = ids\n",
    "                     \n",
    "        self.xTable = np.zeros((self.n,self.m),dtype = int)\n",
    "        self.yTable = np.zeros((self.n,),dtype = int)\n",
    "        \n",
    "        for i in range(self.n):\n",
    "            for j in range(self.m):\n",
    "                self.xTable[i,j] = -1\n",
    "                     \n",
    "        self.__fillXTable()\n",
    "        self.__fillYTable()\n",
    "                     \n",
    "                     \n",
    "    def __fillXTable(self):\n",
    "    \n",
    "        for i in range(self.n):\n",
    "            for j in range(self.m):\n",
    "        \n",
    "                patient = self.ids['patients']['idToName'][i]\n",
    "                snp = self.ids['snps']['idToName'][j]\n",
    "        \n",
    "                self.xTable[i,j] = self.patients[patient].getSnpCode(snp)\n",
    "                     \n",
    "    def __fillYTable(self):\n",
    "    \n",
    "        for i in range(self.n):\n",
    "    \n",
    "            patient = self.ids['patients']['idToName'][i]\n",
    "            self.yTable[i] = self.patients[patient].getCase()\n",
    "        \n",
    "    def getXTable(self):\n",
    "                     \n",
    "        return self.xTable\n",
    "                     \n",
    "    def getYTable(self):\n",
    "                     \n",
    "        return self.yTable\n",
    "    \n",
    "\n",
    "def setIdToName(aList):\n",
    "    \n",
    "    ids = {}\n",
    "    nameToId = {}\n",
    "    idToName = {}\n",
    "    count = 0\n",
    "    \n",
    "    for i in aList:\n",
    "        \n",
    "        nameToId[i] = count\n",
    "        idToName[count] = i\n",
    "        count += 1\n",
    "        \n",
    "    ids['nameToId'] = nameToId\n",
    "    ids['idToName'] = idToName\n",
    "    \n",
    "    return ids\n",
    "\n",
    "\n",
    "def setSnpsCode(patients,chromosomes):\n",
    "    \n",
    "    for i in patients.keys():\n",
    "        patients[i].snpCode(chromosomes)\n",
    "        \n",
    "    return patients\n",
    "    \n",
    "    \n",
    "def mergeXTrainXTestTable(test,train):\n",
    "    \n",
    "    n = len(test) + len(train)\n",
    "    m = len(test.T)\n",
    "    count = 0\n",
    "    \n",
    "    mergeTable = np.zeros((n,m),dtype = int)\n",
    "    \n",
    "    for i in range(len(train)):\n",
    "        for j in range(len(train.T)):\n",
    "            mergeTable[count,j] = train[i,j]\n",
    "        count += 1\n",
    "    \n",
    "    \n",
    "    for i in range(len(test)):\n",
    "        for j in range(len(test.T)):\n",
    "            mergeTable[count,j] = test[i,j]\n",
    "        count += 1\n",
    "\n",
    "        \n",
    "    return mergeTable\n",
    "\n",
    "\n",
    "def mergeYTrainYTestTable(test,train):\n",
    "    \n",
    "    n = len(test) + len(train)\n",
    "   \n",
    "    count = 0\n",
    "    \n",
    "    mergeTable = np.zeros((n,1),dtype = int)\n",
    "    \n",
    "    for i in range(len(train)):\n",
    "        mergeTable[count] = train[i]\n",
    "        count += 1\n",
    "    \n",
    "    for i in range(len(test)):\n",
    "        mergeTable[count] = test[i]\n",
    "        count += 1\n",
    "        \n",
    "    return mergeTable\n",
    "\n",
    "def createAllPatientsStructure(patients,patientsTest,patientsTrain,snps):\n",
    "    \n",
    "    for i in patientsTest.keys():\n",
    "        patients[i]=patientsTest[i]\n",
    "        \n",
    "    for i in patientsTrain.keys():\n",
    "        patients[i]=patientsTrain[i]\n",
    "        \n",
    "    return patients\n",
    "\n",
    "def seperatePatients(patients,allPatients):\n",
    "    \n",
    "    for patient in patients.keys():\n",
    "        \n",
    "        patients[patient] = allPatients[patient]\n",
    "        \n",
    "    return patients"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "path = 'C:\\\\Users\\\\ANTONIS\\\\Desktop\\\\test1\\\\' #bake ton fakelo pou 8a epe3ergas8eis \n",
    "numberOfChromosomes = 14#'ari8mos twn xromoswmatwn'\n",
    "patientsTrain = {}\n",
    "patientsTest = {}\n",
    "allPatients = {}\n",
    "\n",
    "chromosomes = {}\n",
    "\n",
    "read = Read(path,numberOfChromosomes)\n",
    "write = Write(path,numberOfChromosomes)\n",
    "\n",
    "patientsTrain = read.readPatients('phenotype_euro_train_balanced.txt')\n",
    "patientsTest = read.readPatients('phenotype_euro_test_balanced.txt')\n",
    "\n",
    "chromosomes = read.readSnps(\".assoc.fisher\")\n",
    "\n",
    "write.writePatientsList(patientsTrain,'trainPatient.txt')\n",
    "write.writePatientsList(patientsTest,'testPatient.txt')\n",
    "write.writeSnpsList(chromosomes)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# run train_lgen bat and test_leg bat"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "snps = read.getListOfSnps()\n",
    "ids = {} \n",
    "idsTest = {}\n",
    "\n",
    "allPatients = createAllPatientsStructure(allPatients,patientsTest,patientsTrain,snps)\n",
    "\n",
    "if os.path.exists(path + 'snpCode.txt'):\n",
    "    \n",
    "    allPatients = read.readSnpsCode(allPatients)\n",
    "    \n",
    "    patientsTrain = seperatePatients(patientsTrain,allPatients)\n",
    "    patientsTest = seperatePatients(patientsTest,allPatients)\n",
    "    \n",
    "else:\n",
    "    allPatients = read.readLgen(allPatients)\n",
    "    \n",
    "    patientsTrain = seperatePatients(patientsTrain,allPatients)\n",
    "    patientsTest = seperatePatients(patientsTest,allPatients)\n",
    "    \n",
    "    patientsTrain = setSnpsCode(patientsTrain,chromosomes)\n",
    "    patientsTest = setSnpsCode(patientsTest,chromosomes)\n",
    "    \n",
    "\n",
    "if os.path.exists(path + 'TrainIds.txt') and os.path.exists(path + 'SnpsIds.txt'):\n",
    "    ids = read.readIds('Train')\n",
    "else:\n",
    "    ids['patients'] = setIdToName(list(patientsTrain.keys()))\n",
    "    ids['snps'] = setIdToName(snps)\n",
    "\n",
    "if os.path.exists(path + 'Test.txt') and os.path.exists(path + 'SnpsIds.txt'):\n",
    "    idsTest = read.readIds('Test')\n",
    "else:    \n",
    "    idsTest['patients'] = setIdToName(list(patientsTest.keys()))\n",
    "    idsTest['snps'] = ids['snps']\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "trainSet = DataSet(patientsTrain,ids)\n",
    "testSet = DataSet(patientsTest,idsTest)\n",
    "\n",
    "xTraining = trainSet.getXTable()\n",
    "yTraining = trainSet.getYTable()\n",
    "        \n",
    "xTest = testSet.getXTable()\n",
    "yTest = testSet.getYTable()\n",
    "\n",
    "\n",
    "mergeXtable = mergeXTrainXTestTable(xTraining,xTest)\n",
    "mergeYtable = mergeYTrainYTestTable(yTraining,yTest)\n",
    "    \n",
    "write.saveData(ids,patientsTrain,'Train',xTraining,chromosomes)\n",
    "write.saveData(idsTest,patientsTest,'Test',xTest)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "count = 0\n",
    "\n",
    "for patient in patientsTrain.keys():\n",
    "            \n",
    "    if patientsTrain[patient].getSize() == read.getNumberOfSnps():\n",
    "        \n",
    "        count += 1\n",
    "        \n",
    "print(\"countTrain is \",count)\n",
    "print(\"patientsTrain is \",len(patientsTrain.keys())) \n",
    "print(\"snps is Train\",read.getNumberOfSnps())\n",
    "print(\"shape id \",xTraining.shape)\n",
    "print(\"xTraining = \",len(xTraining))\n",
    "print(\"xTraining.T = \",len(xTraining.T))\n",
    "print(\"yTraining = \",len(yTraining))\n",
    "\n",
    "print()\n",
    "\n",
    "count = 0\n",
    "\n",
    "for patient in patientsTest.keys():\n",
    "            \n",
    "    if patientsTest[patient].getSize() == read.getNumberOfSnps():\n",
    "        \n",
    "        count += 1\n",
    "        \n",
    "print(\"count is Test\",count)\n",
    "print(\"patientsTest is \",len(patientsTrain.keys()))\n",
    "print(\"snps is Test\",read.getNumberOfSnps())\n",
    "print(\"shape id Test\",xTest.shape)\n",
    "print(\"xtest = \",len(xTest))\n",
    "print(\"xtest.T = \",len(xTest.T))\n",
    "print(\"ytest = \",len(yTest))\n",
    "\n",
    "print()\n",
    "print(\"mergeXtable = \",len(mergeXtable))\n",
    "print(\"mergeXtable.T = \",len(mergeXtable.T))\n",
    "print(\"mergeYtable = \",len(mergeYtable))\n",
    "print(\"shape id \",mergeXtable.shape)\n",
    "\n",
    "print()\n",
    "\n",
    "\n",
    "print(\"id to name patients\",len(ids['patients']['idToName'].keys()))\n",
    "print(\"name to id patients\",len(ids['patients']['nameToId'].keys()))\n",
    "print()\n",
    "print(\"id to name snps\",len(ids['snps']['idToName'].keys()))\n",
    "print(\"name to id snps\",len(ids['snps']['nameToId'].keys()))\n",
    "\n",
    "print()\n",
    "count0 = 0\n",
    "count1 = 0\n",
    "count2 = 0\n",
    "count = 0\n",
    "\n",
    "for i in range(len(xTraining)):\n",
    "    for j in range(len(xTraining.T)):\n",
    "        if xTraining[i,j] == 2:\n",
    "            count2 += 1\n",
    "                \n",
    "        elif xTraining[i,j] == 1:\n",
    "            count1 += 1\n",
    "                \n",
    "        elif xTraining[i,j] == 0:\n",
    "            \n",
    "            count0 +=1\n",
    "        else:\n",
    "            count += 1\n",
    "            \n",
    "print(\"xTraining\")\n",
    "print(\"count = \",count)\n",
    "print(\"count0 = \",count0)\n",
    "print(\"count1 = \",count1)\n",
    "print(\"coun2 = \",count2)\n",
    "\n",
    "print()\n",
    "\n",
    "count0 = 0\n",
    "count1 = 0\n",
    "count2 = 0\n",
    "count = 0\n",
    "\n",
    "for i in range(len(xTest)):\n",
    "    for j in range(len(xTest.T)):\n",
    "        if xTest[i,j] == 2:\n",
    "            count2 += 1\n",
    "                \n",
    "        elif xTest[i,j] == 1:\n",
    "            count1 += 1\n",
    "                \n",
    "        elif xTest[i,j] == 0:\n",
    "            \n",
    "            count0 +=1\n",
    "        else:\n",
    "            count += 1\n",
    "            \n",
    "print(\"xTest\")\n",
    "print(\"count = \",count)\n",
    "print(\"count0 = \",count0)\n",
    "print(\"count1 = \",count1)\n",
    "print(\"coun2 = \",count2)\n",
    "\n",
    "count0 = 0\n",
    "count1 = 0\n",
    "count2 = 0\n",
    "count = 0\n",
    "\n",
    "for i in range(len(mergeXtable)):\n",
    "    for j in range(len(mergeXtable.T)):\n",
    "        if mergeXtable[i,j] == 2:\n",
    "            count2 += 1\n",
    "                \n",
    "        elif mergeXtable[i,j] == 1:\n",
    "            count1 += 1\n",
    "                \n",
    "        elif mergeXtable[i,j] == 0:\n",
    "            \n",
    "            count0 +=1\n",
    "        else:\n",
    "            count += 1\n",
    "print()\n",
    "            \n",
    "print(\"mergeXtable\")\n",
    "print(\"count = \",count)\n",
    "print(\"count0 = \",count0)\n",
    "print(\"count1 = \",count1)\n",
    "print(\"coun2 = \",count2)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# # RF"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "\n",
    "rfr = RandomForestRegressor(n_estimators = 500, random_state = 2016, verbose = 20,max_depth = None,n_jobs=-1)\n",
    "\n",
    "xtraining = rfr.transform(xTraining)\n",
    "xtest = rfr.transform(xTest)\n",
    "\n",
    "print(xTraining.shape)\n",
    "print(xtraining.shape)\n",
    "print(xtest.shape)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "rfr.fit(xtraining, yTraining)\n",
    "yPredict3 = rfr.predict(xtest)\n",
    "for i in range(len(yPredict3)):\n",
    "    \n",
    "    if yPredict3[i] <= 0.5:\n",
    "        yPredict3[i]=0\n",
    "    else:\n",
    "        yPredict3[i]=1\n",
    "        \n",
    "print(metrics.accuracy_score(yTest,yPredict3))\n",
    "print(metrics.confusion_matrix(yTest,yPredict3))\n",
    "error3 = mean_squared_error(yTest, yPredict3)\n",
    "print(\"error 3 = \",error3)\n",
    "RMSE3 = mean_squared_error(yTest,yPredict3)**0.5\n",
    "print(\"RMSE3 = \",RMSE3)\n",
    "\n",
    "print(\"cros validation = \",cross_val_score(rfr, mergeXtable, mergeYtable, cv=10))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Linear Regression"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "regr = linear_model.LinearRegression()\n",
    "regr.fit(xTraining, yTraining)\n",
    "yPredict1 = regr.predict(xTest)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "count = 0\n",
    "for i in range(len(yPredict1)):\n",
    "    count += yPredict1[i]\n",
    "\n",
    "mo = count / len(yPredict1)\n",
    "\n",
    "for i in range(len(yPredict1)):\n",
    "    if yPredict1[i] <= mo:\n",
    "        yPredict1[i] = 0\n",
    "    else:\n",
    "        yPredict1[i] = 1\n",
    "\n",
    "error1 = mean_squared_error(yTest, yPredict1)\n",
    "print(\"error 1 = \",error1)\n",
    "print(metrics.accuracy_score(yTest,yPredict1))\n",
    "print(metrics.confusion_matrix(yTest,yPredict1))\n",
    "RMSE1 = mean_squared_error(yTest,yPredict1)**0.5\n",
    "print(\"RMSE1 = \",RMSE1)\n",
    "\n",
    "print(\"cros validation = \",cross_val_score(regr, mergeXtable, mergeYtable, cv=10))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# SVM"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "\n",
    "clf = SVC()\n",
    "clf.fit(xTraining, yTraining)\n",
    "yPredict2 = clf.predict(xTest)\n",
    "print(metrics.accuracy_score(yTest,yPredict2))\n",
    "print(metrics.confusion_matrix(yTest,yPredict2))\n",
    "error2 = mean_squared_error(yTest, yPredict2)\n",
    "print(\"error 2 = \",error2)\n",
    "RMSE2 = mean_squared_error(yTest,yPredict2)**0.5\n",
    "print(\"RMSE2 = \",RMSE2)\n",
    "\n",
    "#print(\"cros validation = \",cross_val_score(clf, mergeXtable, mergeYtable, cv=5))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# LINEAR LOGISTIC REGRESSION "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "lr_clf = linear_model.ElasticNetCV()\n",
    "lr_clf.fit(xTraining, yTraining)\n",
    "yPredict4 = lr_clf.predict(xTest)\n",
    "\n",
    "count = 0\n",
    "for i in range(len(yPredict4)):\n",
    "    count += yPredict4[i]\n",
    "\n",
    "mo = count / len(yPredict4)\n",
    "\n",
    "for i in range(len(yPredict4)):\n",
    "    if yPredict4[i] <= mo:\n",
    "        yPredict4[i] = 0\n",
    "    else:\n",
    "        yPredict4[i] = 1\n",
    "\n",
    "print(metrics.accuracy_score(yTest,yPredict4))\n",
    "print(metrics.confusion_matrix(yTest,yPredict4))\n",
    "error4 = mean_squared_error(yTest, yPredict4)\n",
    "print(\"error 4 = \",error4)\n",
    "RMSE4 = mean_squared_error(yTest,yPredict4)**0.5\n",
    "print(\"RMSE4 = \",RMSE4)\n",
    "#print(\"cros validation = \",cross_val_score(lr_clf, mergeXtable, mergeYtable, cv=5))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# OLS"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "ols = sm.OLS(yTraining,xTraining)\n",
    "yPredict5 = ols.fit().predict(xTest)\n",
    "\n",
    "mo5 = 0\n",
    "s5 = 0\n",
    "for i in yPredict5:\n",
    "    s5 += i\n",
    "    \n",
    "mo5 = s5 / len(yPredict5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "for i in range(len(yPredict5)):\n",
    "    \n",
    "    if yPredict5[i] <=mo5:\n",
    "        yPredict5[i]=0\n",
    "    else:\n",
    "        yPredict5[i]=1\n",
    "\n",
    "print(metrics.accuracy_score(yTest,yPredict5))\n",
    "print(metrics.confusion_matrix(yTest,yPredict5))\n",
    "error5 = mean_squared_error(yTest, yPredict5)\n",
    "print(\"error 5 = \",error5)\n",
    "RMSE5 = mean_squared_error(yTest,yPredict5)**0.5\n",
    "print(\"RMSE5 = \",RMSE5)\n",
    "#print(\"cros validation = \",cross_val_score(ols, mergeXtable, mergeYtable, cv=5))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "from sklearn.naive_bayes import GaussianNB\n",
    "gnb = GaussianNB()\n",
    "y_pred = gnb.fit(xTraining, yTraining).predict(xTest)\n",
    "print(metrics.accuracy_score(yTest,y_pred))\n",
    "print(metrics.confusion_matrix(yTest,y_pred))\n",
    "error5 = mean_squared_error(yTest, yPredict5)\n",
    "print(\"error 5 = \",error5)\n",
    "RMSE5 = mean_squared_error(yTest,yPredict5)**0.5\n",
    "print(\"RMSE5 = \",RMSE5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "anaconda-cloud": {},
  "kernelspec": {
   "display_name": "Python [default]",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
