{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "from sklearn import linear_model\n",
    "from sklearn.model_selection import cross_val_score\n",
    "from sklearn.model_selection import cross_val_score\n",
    "from sklearn.metrics import mean_squared_error\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "import os.path\n",
    "import csv\n",
    "from sklearn.svm import SVC\n",
    "\n",
    "class PatientPhenotype:\n",
    "    \n",
    "    def __init__(self, eid, case, sex, yearBirth):\n",
    "        \n",
    "        self.eid = eid\n",
    "        self.case = case\n",
    "        self.sex = sex\n",
    "        self.yearBirth = yearBirth\n",
    "        self.snps = {}\n",
    "        \n",
    "    def getEid(self):\n",
    "        return self.eid\n",
    "     \n",
    "    def getCase(self):\n",
    "        return self.case\n",
    "    \n",
    "    def getSex(self):\n",
    "        return self.sex\n",
    "    \n",
    "    def getYearBirth(self):\n",
    "        return self.yearBirth\n",
    "        \n",
    "    def addSnps(self, snpId, allele1,allele2):\n",
    "        self.snps[snpId] = Snp(snpId,allele1,allele2)\n",
    "        \n",
    "    def snpCode(self,snpId,allele1,allele2):\n",
    "       \n",
    "        self.snps[snpId].setSnpCode(allele1,allele2)\n",
    "        \n",
    "    def getSnpCode(self,snpId):\n",
    "        return self.snps[snpId].getSnpCode()\n",
    "        \n",
    "    def getSize(self):\n",
    "        return len(self.snps)\n",
    "        \n",
    "        \n",
    "class Snp:\n",
    "    \n",
    "    def __init__(self,snpId,allele1,allele2):\n",
    "        \n",
    "        self.snpId = snpId\n",
    "        self.allele1 = allele1\n",
    "        self.allele2 = allele2\n",
    "        self.snpCode = -1\n",
    "        \n",
    "    def getId(self):\n",
    "        \n",
    "        return self.snpId\n",
    "        \n",
    "    def getAllele1(self):\n",
    "        \n",
    "        return self.allele1\n",
    "        \n",
    "    def getAllele2(self):\n",
    "        \n",
    "        return self.allele2\n",
    "        \n",
    "    def setSnpCode(self,allele1,allele2):\n",
    "      \n",
    "        if self.allele1 == allele1 and self.allele2 == allele1:\n",
    "            code = 2\n",
    "           \n",
    "        elif self.allele1 == allele1 and self.allele2 != allele1:\n",
    "            code = 1\n",
    "           \n",
    "        elif self.allele1 != allele1 and self.allele2 == allele1:\n",
    "            code = 1\n",
    "            \n",
    "        elif self.allele1 != allele1 and self.allele2 != allele1:\n",
    "            code = 0\n",
    "            \n",
    "        self.snpCode = code\n",
    "        \n",
    "    def getSnpCode(self):\n",
    "        \n",
    "        return self.snpCode\n",
    "    \n",
    "    \n",
    "class SaveDataSet():\n",
    "    def __init__(self,name,ids,data):\n",
    "        \n",
    "        self.name = name\n",
    "        self.ids = ids\n",
    "        self.data = data\n",
    "        self.__writeLogFile()\n",
    "        self.__writeVariables('patients')\n",
    "        self.__writeVariables('snps')\n",
    "        self.__writeCSV()\n",
    "        \n",
    "    def __writeLogFile(self):\n",
    "        \n",
    "        write = open(self.name+'.log','w')\n",
    "        write.write(\"Log file for the Test Data Set \"+self.name+'\\n')\n",
    "        write.write(self.name + \" has \"+ str(len(self.ids['patients']['nameToId'].keys())) + \" patients \"'\\n')\n",
    "        write.write(self.name + \" has \"+ str(len(self.ids['snps']['nameToId'].keys())) + \" snps \"'\\n')\n",
    "        write.close()\n",
    "        \n",
    "    def __writeVariables(self,kind):\n",
    "        \n",
    "        write = open(self.name+'_'+kind+'.txt','w')\n",
    "        write.write(kind + '\\n')\n",
    "        \n",
    "        for i in range(len(self.ids[kind]['nameToId'].keys())):\n",
    "            name = self.ids[kind]['idToName'][i]\n",
    "            write.write(name + '\\t'+str(self.ids[kind]['nameToId'][name]) + '\\n')\n",
    "        \n",
    "        write.close()\n",
    "        \n",
    "        \n",
    "    def __writeCSV(self):\n",
    "        \n",
    "        try:\n",
    "            \n",
    "            with open(self.name+'.csv', 'w') as csvfile:\n",
    "                writer = csv.writer(csvfile, dialect='excel', quoting=csv.QUOTE_NONNUMERIC)\n",
    "            \n",
    "                line = []\n",
    "                for i in range(len(self.ids['snps']['nameToId'].keys())):\n",
    "                   # if i == 0 :\n",
    "                    #    line += self.ids['snps']['idToName'][i]\n",
    "                    #else:\n",
    "                    line.append(self.ids['snps']['idToName'][i])\n",
    "                writer.writerow(line)\n",
    "            \n",
    "                line = [] \n",
    "                for i in range(len(self.ids['snps']['nameToId'].keys())):\n",
    "                    name = self.ids['snps']['idToName'][i]\n",
    "                 #   if i == 0:\n",
    "                  #      line += str(self.ids['snps']['nameToId'][name])\n",
    "                   # else:\n",
    "                    line.append(str(self.ids['snps']['nameToId'][name]))\n",
    "                writer.writerow(line)\n",
    "            \n",
    "            \n",
    "                for i in range(len(self.ids['patients']['nameToId'].keys())):\n",
    "                    line = []\n",
    "                    for j in range(len(self.ids['snps']['nameToId'].keys())):\n",
    "                        #if j == 0:\n",
    "                        #    line += str(self.data[i,j]) \n",
    "                        #else:\n",
    "                        \n",
    "                        line.append(str(self.data[i,j]))\n",
    "                        \n",
    "                    writer.writerow(line)\n",
    "        finally:\n",
    "            pass   \n",
    "        \n",
    "\n",
    "def setIdToName(aList):\n",
    "    \n",
    "    ids = {}\n",
    "    nameToId = {}\n",
    "    idToName = {}\n",
    "    count = 0\n",
    "    \n",
    "    for i in aList:\n",
    "        \n",
    "        nameToId[i] = count\n",
    "        idToName[count] = i\n",
    "        count += 1\n",
    "        \n",
    "    ids['nameToId'] = nameToId\n",
    "    ids['idToName'] = idToName\n",
    "    \n",
    "    return ids\n",
    "\n",
    "\n",
    "def createDataSet(kind,chromos,patients):\n",
    "    \n",
    "    for i in range(len(chromos.keys())):\n",
    "    \n",
    "        chro = 'chr'+str(i+1)\n",
    "        path = 'C:\\\\Users\\\\ANTONIS\\\\Desktop\\\\test1\\\\'+chro+kind+'.lgen'\n",
    "    \n",
    "        if os.path.exists(path):\n",
    "            f = open(path,'r')\n",
    "    \n",
    "            for line in f:\n",
    "                patients[line.split()[0]].addSnps(line.split()[2], line.split()[3],line.split()[4])\n",
    "    \n",
    "\n",
    "    \n",
    "    for i in range(chromos.keys()):\n",
    "    \n",
    "        chro = 'chr'+str(i+1)\n",
    "    \n",
    "        for snp in chromos[chro].keys():\n",
    "        \n",
    "            allele1 = chromos[chro][snp][0]\n",
    "            allele2 = chromos[chro][snp][1]\n",
    "    \n",
    "            for patient in patients.keys():\n",
    "            \n",
    "                patients[patient].snpCode(snp,allele1,allele2)\n",
    "    \n",
    "    return patients"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "minorSnps = []\n",
    "chromosomes = {}\n",
    "\n",
    "for i in range(14):\n",
    "    \n",
    "    chro = 'chr'+str(i+1)\n",
    "    path = 'C:\\\\Users\\\\ANTONIS\\\\Desktop\\\\test1\\\\'+chro+'.assoc.fisher'\n",
    "    f = open(path,'r')\n",
    "    f.readline()\n",
    "    snps = {}\n",
    "    \n",
    "    for line in f:\n",
    "        alleles = []\n",
    "        minorSnps.append(line.split()[1])\n",
    "        alleles.append(line.split()[3])\n",
    "        alleles.append(line.split()[6])\n",
    "        snps[line.split()[1]] = alleles\n",
    "        \n",
    "    chromosomes[chro] = snps\n",
    "    \n",
    "for i in range(14):\n",
    "    \n",
    "    chro = 'chr'+str(i+1)\n",
    "    path = 'C:\\\\Users\\\\ANTONIS\\\\Desktop\\\\test1\\\\'+chro+'snpList.txt'\n",
    "    write = open(path,'w')\n",
    "    for snp in chromosomes[chro].keys():\n",
    "        write.write(snp + '\\n')\n",
    "            \n",
    "    write.close()\n",
    "    \n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# TRAIN SET CREATION"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "########   TRAIN ##########\n",
    "\n",
    "patientsPhen = {}\n",
    "\n",
    "f = open('C:\\\\Users\\\\Antonis\\\\Documents\\\\GitHub\\\\diplwmatikh\\\\data\\\\phenotype_euro_train.txt','r')\n",
    "f.readline()\n",
    "\n",
    "for line in f:\n",
    "    patientsPhen[line.split()[0]] = PatientPhenotype(line.split()[0],line.split()[3],line.split()[1],line.split()[2])\n",
    "\n",
    "    \n",
    "path = 'C:\\\\Users\\\\ANTONIS\\\\Desktop\\\\test1\\\\'+'trainPatient.txt'\n",
    "write = open(path,'w')\n",
    "for patient in patientsPhen.keys():\n",
    "    write.write(patient + '\\n')\n",
    "            \n",
    "write.close()\n",
    "    \n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# run train_lgen bat"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "ids = {}\n",
    "for i in range(14):\n",
    "    \n",
    "    chro = 'chr'+str(i+1)\n",
    "    path = 'C:\\\\Users\\\\ANTONIS\\\\Desktop\\\\test1\\\\'+chro+'train.lgen'\n",
    "    \n",
    "    if os.path.exists(path):\n",
    "        f = open(path,'r')\n",
    "    \n",
    "        for line in f:\n",
    "            patientsPhen[line.split()[0]].addSnps(line.split()[2], line.split()[3],line.split()[4])\n",
    "    \n",
    "\n",
    "    \n",
    "for i in range(14):\n",
    "    \n",
    "    chro = 'chr'+str(i+1)\n",
    "    \n",
    "    for snp in chromosomes[chro].keys():\n",
    "        \n",
    "        allele1 = chromosomes[chro][snp][0]\n",
    "        allele2 = chromosomes[chro][snp][1]\n",
    "    \n",
    "        for patient in patientsPhen.keys():\n",
    "            \n",
    "            patientsPhen[patient].snpCode(snp,allele1,allele2)\n",
    "            \n",
    "#patientsPhen = createDataSet('train',chromosomes,patientsPhen)\n",
    "\n",
    "count = 0\n",
    "\n",
    "for patient in patientsPhen.keys():\n",
    "            \n",
    "    if patientsPhen[patient].getSize() == len(minorSnps):\n",
    "        \n",
    "        count += 1\n",
    "        \n",
    "print(\"count is \",count)\n",
    "print(\"snps is \",len(minorSnps))\n",
    "\n",
    "\n",
    "ids['patients'] = setIdToName(list(patientsPhen.keys()))\n",
    "ids['snps'] = setIdToName(minorSnps)\n",
    "    \n",
    "xTraining = np.zeros((len(ids['patients']['nameToId'].keys()), len(ids['snps']['nameToId'].keys())),dtype = int)\n",
    "\n",
    "print(\"shape id \",xTraining.shape)\n",
    "print(\"xtraining = \",len(xTraining))\n",
    "print(\"xtraining.T = \",len(xTraining.T))\n",
    "    \n",
    "for i in range(len(xTraining)):\n",
    "    for j in range(len(xTraining.T)):\n",
    "        \n",
    "        patient = ids['patients']['idToName'][i]\n",
    "        snp = ids['snps']['idToName'][j]\n",
    "        \n",
    "        xTraining[i,j] = patientsPhen[patient].getSnpCode(snp)\n",
    "        \n",
    "        \n",
    "yTraining = []\n",
    "\n",
    "for i in range(len(ids['patients']['nameToId'].keys())):\n",
    "    \n",
    "    patient = ids['patients']['idToName'][i]\n",
    "    yTraining.append(int(patientsPhen[patient].getCase()))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "dok = SaveDataSet(\"dok\",ids,xTraining)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# TEST SET CREATION"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "patientsPhenTest = {}\n",
    "\n",
    "f = open('C:\\\\Users\\\\Antonis\\\\Documents\\\\GitHub\\\\diplwmatikh\\\\data\\\\phenotype_euro_test.txt','r')\n",
    "f.readline()\n",
    "for line in f:\n",
    "    patientsPhenTest[line.split()[0]] = PatientPhenotype(line.split()[0],line.split()[3],line.split()[1],line.split()[2])  \n",
    "    \n",
    "    \n",
    "path = 'C:\\\\Users\\\\ANTONIS\\\\Desktop\\\\test1\\\\'+'testPatient.txt'\n",
    "write = open(path,'w')\n",
    "for patient in patientsPhenTest.keys():\n",
    "    write.write(patient + '\\n')\n",
    "            \n",
    "write.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# run test_lgen bat"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "idsTest = {}\n",
    "\n",
    "for i in range(14):\n",
    "    \n",
    "    chro = 'chr'+str(i+1)\n",
    "    path = 'C:\\\\Users\\\\ANTONIS\\\\Desktop\\\\test1\\\\'+chro+'test.lgen'\n",
    "    if os.path.exists(path):\n",
    "        f = open(path,'r')\n",
    "    \n",
    "        for line in f:\n",
    "            patientsPhenTest[line.split()[0]].addSnps(line.split()[2], line.split()[3],line.split()[4])\n",
    "    \n",
    "\n",
    "    \n",
    "for i in range(14):\n",
    "    \n",
    "    chro = 'chr'+str(i+1)\n",
    "    \n",
    "    for snp in chromosomes[chro].keys():\n",
    "        \n",
    "        allele1 = chromosomes[chro][snp][0]\n",
    "        allele2 = chromosomes[chro][snp][1]\n",
    "        \n",
    "        for patient in patientsPhenTest.keys():\n",
    "            \n",
    "            patientsPhenTest[patient].snpCode(snp,allele1,allele2)\n",
    "            \n",
    "#patientsPhenTest = createDataSet('test',chromosomes,patientsPhenTest)            \n",
    "\n",
    "    \n",
    "count = 0\n",
    "\n",
    "for patient in patientsPhenTest.keys():\n",
    "            \n",
    "    if patientsPhenTest[patient].getSize() == len(minorSnps):\n",
    "        \n",
    "        count += 1\n",
    "        \n",
    "print(\"count is Test\",count)\n",
    "print(\"snps is Test\",len(minorSnps))\n",
    "\n",
    "\n",
    "idsTest['patients'] = setIdToName(list(patientsPhenTest.keys()))\n",
    "idsTest['snps'] = setIdToName(minorSnps)\n",
    "    \n",
    "xTest = np.zeros((len(idsTest['patients']['nameToId'].keys()), len(idsTest['snps']['nameToId'].keys())),dtype = int)\n",
    "\n",
    "print(\"shape id Test\",xTest.shape)\n",
    "print(\"xtraining = \",len(xTest))\n",
    "print(\"xtest.T = \",len(xTest.T))\n",
    "    \n",
    "for i in range(len(xTest)):\n",
    "    for j in range(len(xTest.T)):\n",
    "        \n",
    "        patient = idsTest['patients']['idToName'][i]\n",
    "        snp = idsTest['snps']['idToName'][j]\n",
    "        \n",
    "        xTest[i,j] = patientsPhenTest[patient].getSnpCode(snp)\n",
    "        \n",
    "        \n",
    "yTest = []\n",
    "\n",
    "for i in range(len(idsTest['patients']['nameToId'].keys())):\n",
    "    patient = idsTest['patients']['idToName'][i]\n",
    "    yTest.append(int(patientsPhenTest[patient].getCase()))\n",
    "\n",
    "yTest = np.array(yTest)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Linear Regression"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "from sklearn import linear_model\n",
    "from sklearn.model_selection import cross_val_score\n",
    "from sklearn.metrics import mean_squared_error\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "regr = linear_model.LinearRegression()\n",
    "regr.fit(xTraining, yTraining)\n",
    "yPredict1 = regr.predict(xTest)\n",
    "error1 = mean_squared_error(yTest, yPredict1)\n",
    "print(\"error 1 = \",error1)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Cross validation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "from sklearn.svm import SVC\n",
    "clf = SVC()\n",
    "clf.fit(xTraining, yTraining)\n",
    "yPredict2 = clf.predict(xTest)\n",
    "error1 = mean_squared_error(yTest, yPredict2)\n",
    "print(\"error 1 = \",error1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "count0 =0 \n",
    "count1= 0\n",
    "count2 = 0\n",
    "count = 0\n",
    "\n",
    "for i in range(len(xTest)):\n",
    "    for j in range(len(xTest.T)):\n",
    "        if xTest[i,j] == 0:\n",
    "            count0 += 1\n",
    "        elif xTest[i,j] == 1:\n",
    "            count1 += 1\n",
    "        elif xTest[i,j] == 2:\n",
    "            count2 += 1\n",
    "        else:\n",
    "            count += 1\n",
    "print(\"count0 = \",count0)\n",
    "print(\"count1 = \",count1)\n",
    "print(\"count2 = \",count2)\n",
    "print(\"count = \",count)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "count0 =0 \n",
    "count1= 0\n",
    "count2 = 0\n",
    "count = 0\n",
    "\n",
    "for i in range(len(xTraining)):\n",
    "    for j in range(len(xTraining.T)):\n",
    "        if xTraining[i,j] == 0:\n",
    "            count0 += 1\n",
    "        elif xTraining[i,j] == 1:\n",
    "            count1 += 1\n",
    "        elif xTraining[i,j] == 2:\n",
    "            count2 += 1\n",
    "        else:\n",
    "            count += 1\n",
    "print(\"count0 = \",count0)\n",
    "print(\"count1 = \",count1)\n",
    "print(\"count2 = \",count2)\n",
    "print(\"count = \",count)"
   ]
  }
 ],
 "metadata": {
  "anaconda-cloud": {},
  "kernelspec": {
   "display_name": "Python [conda root]",
   "language": "python",
   "name": "conda-root-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
